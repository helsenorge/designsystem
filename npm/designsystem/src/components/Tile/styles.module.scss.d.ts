export type Styles = {
  tile: string;
  tile__description: string;
  tile__title: string;
  'tile__title--compact': string;
  'tile__title--highlighted': string;
  'tile--compact': string;
  'tile--fixed': string;
  'tile--highlighted': string;
  'title-wrapper': string;
  'title-wrapper--compact': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
