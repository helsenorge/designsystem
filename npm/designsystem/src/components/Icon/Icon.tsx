import React from 'react';

import classNames from 'classnames';

import { AnalyticsId, IconSize } from '../../constants';
import { useUuid } from '../../hooks/useUuid';

export type SvgIcon = React.FC<SvgPathProps>;

export interface BaseIconProps {
  /* aria-label for the <svg> element. Used as <title> tag. */
  ariaLabel?: string;
  /* Changes the size of the icon. */
  size?: number;
  /* Changes the color of the icon. */
  color?: string;
  /* Changes the hover color of the icon. */
  hoverColor?: string;
  /* Adds custom classes to the element. */
  className?: string;
  /* Swaps the displayed icon to the hover version and changes its color. */
  isHovered?: boolean;
  /** Sets the data-testid attribute. */
  testId?: string;
}

export interface IconProps extends BaseIconProps {
  /* Sets which icon should be displayed. */
  svgIcon: SvgIcon;
}

export interface SvgPathProps {
  size: IconSize;
  isHovered: boolean;
}

interface IconConfig {
  size: IconSize;
  isHovered: boolean;
  normal: React.ReactElement;
  normalHover: React.ReactElement;
  xSmall?: React.ReactElement;
  xSmallHover?: React.ReactElement;
  xxSmall?: React.ReactElement;
  xxSmallHover?: React.ReactElement;
}

export const getIcon = ({
  size,
  isHovered,
  normal,
  normalHover,
  xSmall,
  xSmallHover,
  xxSmall,
  xxSmallHover,
}: IconConfig): React.ReactElement => {
  if (size <= IconSize.XXSmall && xxSmall && xxSmallHover) {
    return isHovered ? xxSmallHover : xxSmall;
  }
  if (size <= IconSize.XSmall && xSmall && xSmallHover) {
    return isHovered ? xSmallHover : xSmall;
  }

  return isHovered ? normalHover : normal;
};

export const Icon = React.forwardRef((props: IconProps, ref: React.ForwardedRef<SVGSVGElement>) => {
  const {
    svgIcon,
    ariaLabel,
    className = '',
    size = IconSize.Small,
    color = 'black',
    hoverColor = color || 'black',
    isHovered = false,
    testId,
    ...other
  } = props;

  const svgRaw = React.createElement(svgIcon, {
    size,
    isHovered,
  });

  const titleId = useUuid();
  const svgColor = isHovered ? hoverColor : color;

  return (
    <svg
      data-testid={testId}
      data-analyticsid={AnalyticsId.Icon}
      ref={ref}
      className={classNames(`hnds-style-icon`, className)}
      role={ariaLabel ? 'img' : 'presentation'}
      aria-labelledby={ariaLabel ? titleId : undefined}
      focusable={false}
      aria-hidden={ariaLabel ? undefined : true}
      viewBox="0 0 48 48"
      style={{ minWidth: size, minHeight: size }}
      width={size}
      height={size}
      fill={svgColor}
      color={svgColor}
      {...other}
    >
      {ariaLabel && <title id={titleId}>{ariaLabel}</title>}
      {svgRaw}
    </svg>
  );
});

Icon.displayName = 'Icon';

export default Icon;
