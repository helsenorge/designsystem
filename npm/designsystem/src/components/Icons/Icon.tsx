import React from 'react';
import { AnalyticsId, IconSize } from '../../constants';

type SvgIcon = React.FC<SvgPathProps>;

interface IconProps {
  /* Sets which icon should be displayed. */
  svgIcon: SvgIcon;
  /* aria-label for the <svg> element. Used as <title> tag if id is set */
  ariaLabel?: string;
  /* id to */
  id?: string;
  /* Changes the size of the icon. */
  size?: number;
  /* Changes the color of the icon. */
  color?: string;
  /* Changes the hover color of the icon. */
  hoverColor?: string;
  /* Adds custom classes to the element. */
  className?: string;
  /* Use a custom hover icon. */
  hoverType?: SvgIcon;
  /* Swaps the displayed icon to the hover version and changes its color. */
  isHovered?: boolean;
  /** Sets the data-testid attribute. */
  testId?: string;
}

interface SvgPathProps {
  size: IconSize;
  isHovered: boolean;
}

export const getIcon = (
  size: IconSize,
  isHovered: boolean,
  normal: JSX.Element,
  normalHover: JSX.Element,
  small?: JSX.Element,
  smallHover?: JSX.Element,
  xxSmall?: JSX.Element,
  xxSmallHover?: JSX.Element
): JSX.Element => {
  if (size === IconSize.XSmall && small && smallHover) {
    return isHovered ? smallHover : small;
  }

  if (size === IconSize.XXSmall && xxSmall && xxSmallHover) {
    return isHovered ? xxSmallHover : xxSmall;
  }

  return isHovered ? normalHover : normal;
};

const Icon = React.forwardRef((props: IconProps, ref: React.ForwardedRef<SVGSVGElement>) => {
  const {
    svgIcon,
    ariaLabel,
    id,
    className = '',
    size = IconSize.Small,
    color = 'black',
    hoverColor = color || 'black',
    isHovered = false,
    testId,
    ...other
  } = props;

  const svgRaw = React.createElement(svgIcon, {
    size,
    isHovered,
  });

  const labelledby = id && ariaLabel ? `title-${id}` : undefined;

  return (
    <svg
      id={id}
      data-testid={testId}
      data-analyticsid={AnalyticsId.Icon}
      ref={ref as React.RefObject<SVGSVGElement>}
      className={`hnds-style-icon ${className}`}
      role="img"
      aria-label={!id && ariaLabel ? ariaLabel : undefined}
      aria-labelledby={labelledby}
      aria-hidden="true"
      viewBox="0 0 48 48"
      style={{ minWidth: size, minHeight: size }}
      width={size}
      height={size}
      fill={isHovered ? hoverColor : color}
      {...other}
    >
      {labelledby && <title id={labelledby}>{ariaLabel}</title>}
      {svgRaw}
    </svg>
  );
});

Icon.displayName = 'Icon';

export { Icon, IconProps, SvgPathProps, SvgIcon };
