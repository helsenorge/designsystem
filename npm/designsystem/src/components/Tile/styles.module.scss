@use 'sass:map';
@import '../../scss/palette';
@import '../../scss/spacers';
@import '../../scss/breakpoints';

@mixin tile-title {
  overflow-wrap: break-word;
  word-wrap: break-word;
  word-break: break-word;
  font-size: 1.5rem;
  text-align: left;
  line-height: 1.75rem;
  font-weight: 600;
  margin: 0;
  margin-left: 0.5rem;

  @media (min-width: map.get($grid-breakpoints, sm)) {
    font-size: 1.25rem;
    margin-top: 1rem;
  }
}

.tile {
  // Referanse for grandparent selection i underklassene lenger nede
  $tile: &;

  display: flex;
  width: inherit;
  padding: 1.5rem 2rem 2rem 1rem;
  flex-direction: column;
  justify-content: flex-start;
  background-color: $white;
  color: $black;
  text-decoration: none;
  cursor: pointer;
  outline: none;

  @media print {
    border: 1px solid $black;
  }

  &--fixed {
    min-width: 15rem;
    max-width: 20rem;
  }

  &--compact {
    padding: 1.5rem 1rem;
  }

  svg {
    flex-shrink: 0;
  }

  &:hover,
  &:active,
  &:focus {
    background-color: $neutral100;
  }

  &:focus {
    box-shadow: 0 0 0 6px $neutral600;
  }

  &--highlighted {
    background-color: $blueberry500;
    color: $white;

    &:hover,
    &:active,
    &:focus {
      background-color: $blueberry700;
    }

    &:focus {
      box-shadow: 0 0 0 6px $blueberry700;
    }
  }

  &__title {
    @include tile-title;

    &--compact {
      margin: 0 0 0 1.5rem;
    }
  }

  #{$tile}:focus & {
    &__title {
      &--highlighted {
        text-decoration: underline;
      }
    }
  }

  &__description {
    font-size: 1.125rem;
    line-height: 1.75rem;
    margin: 0 0 0 0.5rem;
  }
}

.title-wrapper {
  display: flex;
  flex-direction: row;
  align-items: center;
  padding-bottom: 0.5rem;

  @media (min-width: map.get($grid-breakpoints, sm)) {
    flex-direction: column;
    align-items: start;

    @media print {
      flex-direction: row;
    }
  }

  &--compact {
    flex-direction: row;
    align-items: center;
    padding-right: 1rem;
    padding-bottom: 0;

    svg {
      flex-shrink: 0;
    }
  }
}
