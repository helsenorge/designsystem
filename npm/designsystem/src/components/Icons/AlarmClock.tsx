import React from 'react';
import { getIcon, SvgPathProps } from './Icon';

const AlarmClock: React.FC<SvgPathProps> = ({ size, isHovered }: SvgPathProps): JSX.Element => {
  const normal = (
    <g fillRule="evenodd">
      <path d="M24.65 25.937h-4.88v-1.3h3.58v-4.862h1.3zM16.272 36.651l-.954-.883 2.632-2.845.954.883zM31.728 36.651l-2.632-2.845.954-.883 2.632 2.845z" />
      <path d="M24 14.602c-5.447 0-9.879 4.432-9.879 9.88 0 5.446 4.432 9.877 9.879 9.877 5.447 0 9.878-4.43 9.878-9.878 0-5.447-4.431-9.879-9.878-9.879m0 21.057c-6.164 0-11.179-5.015-11.179-11.178 0-6.164 5.015-11.179 11.179-11.179 6.164 0 11.178 5.015 11.178 11.18 0 6.162-5.014 11.177-11.178 11.177" />
      <path d="M16.287 12.44h-.005a3.418 3.418 0 00-2.435 1.013 3.435 3.435 0 00.557 5.306 11.19 11.19 0 014.936-4.43 3.407 3.407 0 00-3.053-1.89zm-1.33 8.03l-.556-.241a4.74 4.74 0 01-2.859-4.336 4.707 4.707 0 011.384-3.357 4.715 4.715 0 013.354-1.396h.007c2.08 0 3.892 1.344 4.513 3.346l.18.579-.565.219a9.885 9.885 0 00-5.178 4.649l-.28.538zm13.704-6.14a11.179 11.179 0 014.935 4.428 3.438 3.438 0 001.563-2.879 3.45 3.45 0 00-3.446-3.44h-.006a3.407 3.407 0 00-3.046 1.89zm4.382 6.14l-.28-.537a9.884 9.884 0 00-5.178-4.649l-.563-.22.179-.578a4.713 4.713 0 014.504-3.346h.008a4.75 4.75 0 014.745 4.737 4.738 4.738 0 01-2.86 4.35l-.555.242z" />
    </g>
  );

  const normalHover = (
    <g fillRule="evenodd">
      <path d="M24.65 25.937h-4.88v-1.3h3.58v-4.862h1.3zM16.272 36.651l-.954-.883 2.632-2.845.954.883zM31.728 36.651l-2.632-2.845.954-.883 2.632 2.845z" />
      <path d="M24 14.602c-5.447 0-9.88 4.432-9.88 9.88 0 5.446 4.433 9.877 9.88 9.877 5.447 0 9.878-4.43 9.878-9.878 0-5.447-4.431-9.879-9.878-9.879m0 21.057c-6.164 0-11.18-5.015-11.18-11.178 0-6.164 5.016-11.179 11.18-11.179s11.178 5.015 11.178 11.18c0 6.162-5.014 11.177-11.178 11.177" />
      <path d="M28.66 14.329a11.184 11.184 0 014.936 4.43 3.437 3.437 0 00-1.878-6.319h-.006a3.407 3.407 0 00-3.052 1.889zm4.383 6.142l-.28-.538a9.88 9.88 0 00-5.179-4.649l-.564-.22.18-.578c.621-2.002 2.434-3.346 4.512-3.346h.008a4.715 4.715 0 013.354 1.396 4.71 4.71 0 011.384 3.357 4.742 4.742 0 01-2.859 4.336l-.556.242zm7.718-7.195l-1.3-.008c.014-2.44-1.794-5.082-4.737-5.1l.008-1.3c3.746.022 6.047 3.343 6.03 6.408" />
      <path d="M43.778 13.225l-1.3-.008c.011-1.962-.72-3.954-2.006-5.465-.99-1.163-2.744-2.552-5.477-2.568l.008-1.299c2.516.014 4.81 1.088 6.46 3.024 1.484 1.744 2.328 4.046 2.315 6.316m-27.491-.785h-.005a3.418 3.418 0 00-2.435 1.013 3.435 3.435 0 00.557 5.306 11.179 11.179 0 014.936-4.43 3.407 3.407 0 00-3.053-1.89zm-1.33 8.03l-.556-.241a4.74 4.74 0 01-2.86-4.336 4.707 4.707 0 011.385-3.357 4.715 4.715 0 013.354-1.396h.007c2.079 0 3.892 1.344 4.513 3.346l.179.579-.564.219a9.882 9.882 0 00-5.178 4.649l-.28.538zm-7.718-7.194c-.018-3.065 2.283-6.386 6.03-6.407l.008 1.3c-2.944.017-4.752 2.66-4.738 5.099l-1.3.008z" />
      <path d="M4.254 13.006c-.025-4.378 3.256-9.122 8.6-9.152l.007 1.3c-4.54.025-7.328 4.092-7.307 7.844l-1.3.008z" />
    </g>
  );

  const small = (
    <path
      fillRule="evenodd"
      d="M23.869 33.994c-5.32 0-9.646-4.326-9.646-9.644a9.588 9.588 0 011.98-5.83c.03-.04.06-.082.09-.12a9.601 9.601 0 011.07-1.15l.13-.118a9.554 9.554 0 011.375-1.013 9.573 9.573 0 015-1.413 9.577 9.577 0 014.997 1.41l.108.066a9.63 9.63 0 011.397 1.066c.387.354.747.737 1.072 1.15.032.04.061.082.092.122a9.576 9.576 0 011.98 5.83c0 5.318-4.327 9.644-9.645 9.644zM12.945 15.76a3.187 3.187 0 01.936-2.272 3.19 3.19 0 012.27-.944h.006c1.137 0 2.158.599 2.726 1.549-.028.012-.05.031-.077.044a11.47 11.47 0 00-2.445 1.64c-.08.07-.156.14-.233.21-.32.297-.628.61-.913.942-.03.035-.062.065-.091.1-.312.372-.596.769-.859 1.18-.019.028-.043.053-.062.084a3.196 3.196 0 01-1.258-2.533zm18.632-3.216h.005c.855 0 1.66.332 2.266.936a3.196 3.196 0 01-.315 4.811c-.019-.03-.044-.056-.063-.087-.264-.41-.548-.806-.86-1.178l-.082-.09c-.29-.337-.6-.653-.925-.954l-.223-.203a11.462 11.462 0 00-2.445-1.64c-.027-.013-.053-.032-.08-.046a3.165 3.165 0 012.722-1.549zm3.705 11.807c0-1.593-.33-3.11-.923-4.488a4.95 4.95 0 00.738-7.635 4.947 4.947 0 00-3.515-1.452h-.008a4.926 4.926 0 00-4.37 2.661 11.37 11.37 0 00-6.67 0 4.924 4.924 0 00-4.377-2.661h-.009a4.945 4.945 0 00-3.519 1.464 4.94 4.94 0 00-1.454 3.523 4.968 4.968 0 002.202 4.101 11.33 11.33 0 00-.922 4.487c0 3.686 1.763 6.96 4.484 9.048l-1.925 2.08 1.299 1.2 2.122-2.294a11.34 11.34 0 005.434 1.379c1.965 0 3.817-.5 5.434-1.379l2.12 2.294 1.3-1.2-1.926-2.08c2.72-2.088 4.485-5.362 4.485-9.048zm-12.297-.078h-3.346v1.768h5.114v-6.395h-1.768v4.627z"
    />
  );

  const smallHover = (
    <path
      fillRule="evenodd"
      d="M33.534 18.292c-.021-.033-.048-.062-.07-.095-.26-.404-.539-.795-.847-1.162-.037-.043-.077-.082-.114-.125a11.665 11.665 0 00-1.134-1.14 11.395 11.395 0 00-2.434-1.632c-.028-.013-.053-.032-.08-.046a3.168 3.168 0 012.725-1.549h.006a3.192 3.192 0 012.27.944 3.198 3.198 0 01-.322 4.805zm-9.666 15.702c-5.318 0-9.644-4.326-9.644-9.644a9.578 9.578 0 011.98-5.83l.09-.121a9.623 9.623 0 011.072-1.15 9.615 9.615 0 011.5-1.128 9.58 9.58 0 015.002-1.415 9.58 9.58 0 014.996 1.41l.111.067c.447.28.868.598 1.262.944l.142.129a9.7 9.7 0 011.058 1.137c.036.045.07.093.105.139a9.573 9.573 0 011.97 5.818c0 5.318-4.326 9.644-9.644 9.644zM12.945 15.759a3.213 3.213 0 013.204-3.216h.007c1.138 0 2.158.599 2.727 1.549-.026.012-.048.03-.075.043a11.458 11.458 0 00-3.595 2.796c-.028.033-.06.063-.087.096-.313.373-.597.77-.861 1.18-.02.03-.043.056-.062.085a3.196 3.196 0 01-1.258-2.533zm23.616.003a4.947 4.947 0 00-1.453-3.523 4.945 4.945 0 00-3.52-1.464h-.008c-1.87 0-3.53 1.04-4.378 2.661a11.355 11.355 0 00-3.334-.499c-1.16 0-2.278.176-3.333.5a4.926 4.926 0 00-4.38-2.662h-.008a4.986 4.986 0 00-4.97 4.987 4.963 4.963 0 002.2 4.101 11.352 11.352 0 00-.921 4.487c0 3.686 1.763 6.959 4.483 9.048l-1.925 2.079 1.297 1.202 2.123-2.295c1.617.88 3.469 1.38 5.434 1.38 1.966 0 3.816-.5 5.433-1.38l2.124 2.295 1.297-1.202-1.925-2.08c2.72-2.088 4.484-5.361 4.484-9.047 0-1.593-.33-3.109-.922-4.487a4.963 4.963 0 002.202-4.101zm2.661-7.097c-1.18-1.383-2.82-2.152-4.62-2.162l-.01 1.768c1.638.01 2.689.844 3.284 1.541a5.158 5.158 0 011.218 3.323l1.77.01a6.96 6.96 0 00-1.642-4.48zM13.137 6.503c-3.892.022-6.282 3.466-6.264 6.643l1.768-.01c-.013-2.327 1.707-4.849 4.504-4.865l-.008-1.768zm9.846 17.768h-3.345v1.769h5.115v-6.395h-1.77v4.626z"
    />
  );

  return getIcon(size, isHovered, normal, normalHover, small, smallHover);
};

export default AlarmClock;
