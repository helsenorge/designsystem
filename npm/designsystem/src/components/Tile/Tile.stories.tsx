import React from 'react';

import { action } from '@storybook/addon-actions';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import Tile from './Tile';
import GridExample from '../GridExample';
import Icon, { IconSize } from '../Icons';
import AlarmClock from '../Icons/AlarmClock';

export default {
  title: 'Components/Tile',
  component: Tile,
  parameters: {
    docs: {
      description: {
        component: 'Som innbygger vil jeg kunne vise knapp og link innhold i ulike layouts og formater.',
      },
    },
  },
  argTypes: {
    fixed: {
      control: 'boolean',
      defaultValue: false,
    },
    highlighted: {
      control: 'boolean',
      defaultValue: false,
    },
    href: {
      control: 'text',
      defaultValue: 'https://www.helsenorge.no',
    },
    description: {
      control: 'text',
      defaultValue:
        'Førstehjelp de første minuttene etter at en akutt sykdom eller skade har oppstått er livsviktig og minsker risikoen for langtidsskader.',
    },
  },
} as ComponentMeta<typeof Tile>;

export const Default: ComponentStory<typeof Tile> = (args: any) => (
  <GridExample>
    <Tile
      {...args}
      icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
      title={<Tile.Title htmlMarkup={'h1'}>{'Inbox'}</Tile.Title>}
    />
  </GridExample>
);

export const External: ComponentStory<typeof Tile> = (args: any) => (
  <GridExample>
    <Tile
      {...args}
      icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
      href={'https://www.helsenorge.no'}
      title={<Tile.Title htmlMarkup={'h1'}>{'External'}</Tile.Title>}
    />
  </GridExample>
);

export const WrappedContainer: ComponentStory<typeof Tile> = (args: any) => (
  <GridExample>
    <div
      style={{
        display: 'grid',
        gridTemplateColumns: 'repeat(3, 33.3%)',
        gridGap: '1.5rem',
      }}
    >
      <Tile
        href="https://www.helsenorge.no"
        icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
        title={<Tile.Title>{'Hva skal du si når du ringer 113?'}</Tile.Title>}
        description="Hva du heter.
      Hvor du ringer fra.
      Telefonnummer du ringer fra.
      Beskriv situasjonen og symptomer."
      />
      <Tile
        href="https://www.helsenorge.no"
        icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
        title={<Tile.Title>{'Hva skal du si når du ringer 113?'}</Tile.Title>}
        description="Hva du heter.
      Hvor du ringer fra.
      Telefonnummer du ringer fra.
      Beskriv situasjonen og symptomer."
      />
      <Tile
        href="https://www.helsenorge.no"
        icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
        title={<Tile.Title>{'Hva skal du si når du ringer 113?'}</Tile.Title>}
        description="Hva du heter.
      Hvor du ringer fra.
      Telefonnummer du ringer fra.
      Beskriv situasjonen og symptomer."
      />
      <Tile
        href="https://www.helsenorge.no"
        icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
        title={<Tile.Title>{'Hva skal du si når du ringer 113?'}</Tile.Title>}
        description="Hva du heter.
      Hvor du ringer fra.
      Telefonnummer du ringer fra.
      Beskriv situasjonen og symptomer."
      />
      <Tile
        href="https://www.helsenorge.no"
        icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
        title={<Tile.Title>{'Hva skal du si når du ringer 113?'}</Tile.Title>}
        description="Hva du heter.
      Hvor du ringer fra.
      Telefonnummer du ringer fra.
      Beskriv situasjonen og symptomer."
      />
      <Tile
        href="https://www.helsenorge.no"
        icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
        title={<Tile.Title>{'Hva skal du si når du ringer 113?'}</Tile.Title>}
        description="Hva du heter.
      Hvor du ringer fra.
      Telefonnummer du ringer fra.
      Beskriv situasjonen og symptomer."
      />
      <Tile
        href="https://www.helsenorge.no"
        icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
        title={<Tile.Title>{'Europeisk helsetrygdekort (nav)'}</Tile.Title>}
      />
      <Tile
        href="https://www.helsenorge.no"
        icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
        title={<Tile.Title>{'Europeisk helsetrygdekort (nav)'}</Tile.Title>}
      />
    </div>
  </GridExample>
);

export const WithOnClick: ComponentStory<typeof Tile> = (args: any) => (
  <GridExample>
    <Tile
      {...args}
      onClick={action('Tile clicked!')}
      icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
      title={<Tile.Title htmlMarkup={'h1'}>{'External'}</Tile.Title>}
    />
  </GridExample>
);
