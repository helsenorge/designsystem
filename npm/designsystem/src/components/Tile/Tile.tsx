import React from 'react';

import classNames from 'classnames';

import { TitleTags } from './../Title/Title';
import { AnalyticsId } from '../../constants';
import { useHover } from '../../hooks/useHover';
import { mergeRefs } from '../../utils/refs';
import { IconSize } from '../Icons';

import tileStyles from './styles.module.scss';

interface TileProps extends Pick<React.AnchorHTMLAttributes<HTMLAnchorElement>, 'href' | 'target' | 'onClick' | 'rel'> {
  /** Adds custom classes to the element. */
  className?: string;
  /**	Sets the icon to be displayed inside the tile. */
  icon: React.ReactElement;
  /**	Sets the title to be displayed inside the tile. */
  title: React.ReactElement;
  /** Toggles the highlighted style of the tile. */
  highlighted?: boolean;
  /** Sets the description to be displayed inside the tile. */
  description?: string;
  /** Sets a fixed max and min width for the tile. */
  fixed?: boolean;
  /** Sets the data-testid attribute. */
  testId?: string;
}

interface TileTitleProps {
  children: React.ReactNode;
  className?: string;
  htmlMarkup?: TitleTags;
  highlighted?: boolean;
  compact?: boolean;
}

export interface TileCompound extends React.ForwardRefExoticComponent<TileProps & React.RefAttributes<HTMLAnchorElement>> {
  Title: React.ForwardRefExoticComponent<TileTitleProps & React.RefAttributes<HTMLHeadingElement>>;
}

const Title = React.forwardRef<HTMLHeadingElement, TileTitleProps>((props, ref) => {
  const { children, className, htmlMarkup = 'span', highlighted, compact } = props;
  const titleClasses = classNames(
    tileStyles['tile__title'],
    {
      [tileStyles['tile__title--highlighted']]: highlighted,
      [tileStyles['tile__title--compact']]: compact,
    },
    className
  );
  const CustomTag = htmlMarkup;

  return (
    <CustomTag className={titleClasses} ref={ref}>
      {children}
    </CustomTag>
  );
});

Title.displayName = 'Title';

export const Tile = React.forwardRef<HTMLAnchorElement, TileProps>((props, ref) => {
  const { icon, title, className = '', description, fixed = false, highlighted = false, testId, target, rel, href, onClick } = props;
  const { hoverRef, isHovered } = useHover<HTMLAnchorElement>();
  const compact = !description;
  const tileClasses = classNames(
    tileStyles.tile,
    {
      [tileStyles['tile--fixed']]: fixed,
      [tileStyles['tile--compact']]: compact,
      [tileStyles['tile--highlighted']]: highlighted,
    },
    className
  );
  const tileTitleWrapperClasses = classNames(tileStyles['title-wrapper'], {
    [tileStyles['title-wrapper--compact']]: compact,
  });

  return (
    <a
      ref={mergeRefs([ref, hoverRef])}
      href={href}
      target={target}
      rel={target === '_blank' ? 'noopener noreferrer' : rel}
      className={tileClasses}
      data-testid={testId}
      data-analyticsid={AnalyticsId.Tile}
      onClick={onClick}
    >
      <div className={tileTitleWrapperClasses}>
        {React.cloneElement(icon, { size: IconSize.Medium, isHovered, color: highlighted ? 'white' : 'black' })}
        {React.cloneElement(title, { highlighted: highlighted, compact: compact })}
      </div>
      {description && <p className={tileStyles.tile__description}>{description}</p>}
    </a>
  );
}) as TileCompound;

Tile.displayName = 'Tile';
Tile.Title = Title;

export default Tile;
