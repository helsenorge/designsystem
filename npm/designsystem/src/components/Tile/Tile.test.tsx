import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Tile from './Tile';
import { IconSize } from '../../constants';
import Icon from '../Icons';
import AlarmClock from '../Icons/AlarmClock';

test('displays tile', (): void => {
  const { container } = render(
    <Tile
      icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
      title={<Tile.Title>{'Title'}</Tile.Title>}
      description="Description"
    />
  );
  expect(container).toMatchSnapshot();
});

describe('Gitt at Tile skal vises', (): void => {
  describe('Når testId-prop er satt', (): void => {
    test('Så kan komponenten finnes ved hjelp av testId', (): void => {
      render(
        <Tile
          icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
          title={<Tile.Title>{'Title'}</Tile.Title>}
          description="Description"
          testId="bare-tester"
        />
      );

      const component = screen.getByTestId('bare-tester');
      expect(component).toBeVisible();
    });
  });

  describe('Når target=_blank', (): void => {
    test('Skal Tile ha riktig rel-attributt', (): void => {
      render(
        <Tile
          icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
          title={<Tile.Title>{'Title'}</Tile.Title>}
          description="Description"
          target="_blank"
          testId="tester-rel"
        />
      );

      const component = screen.getByTestId('tester-rel');
      expect(component).toBeVisible();
      expect(component).toHaveAttribute('target', '_blank');
      expect(component).toHaveAttribute('rel', 'noopener noreferrer');
    });
  });
  describe('Når onClick er satt', () => {
    test('Skal kalles onClick-handler når man klikker på knappen', async () => {
      const onClickMock = jest.fn();
      render(
        <Tile
          icon={<Icon size={IconSize.Medium} svgIcon={AlarmClock} />}
          title={<Tile.Title>{'Lenketekst'}</Tile.Title>}
          onClick={onClickMock}
          href="#"
        />
      );

      const tile = screen.getByRole('link', { name: 'Lenketekst' });
      await userEvent.click(tile);
      expect(onClickMock).toHaveBeenCalledTimes(1);
    });
  });
});
