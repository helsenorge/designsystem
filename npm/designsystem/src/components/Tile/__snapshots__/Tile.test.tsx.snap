// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays tile 1`] = `
<div>
  <a
    class="tile"
    data-analyticsid="tile"
  >
    <div
      class="title-wrapper"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon"
        data-analyticsid="icon"
        fill="black"
        focusable="false"
        height="64"
        role="presentation"
        style="min-width: 64px; min-height: 64px;"
        viewBox="0 0 48 48"
        width="64"
      >
        <g
          fill-rule="evenodd"
        >
          <path
            d="M24.65 25.937h-4.88v-1.3h3.58v-4.862h1.3zM16.272 36.651l-.954-.883 2.632-2.845.954.883zM31.728 36.651l-2.632-2.845.954-.883 2.632 2.845z"
          />
          <path
            d="M24 14.602c-5.447 0-9.879 4.432-9.879 9.88 0 5.446 4.432 9.877 9.879 9.877 5.447 0 9.878-4.43 9.878-9.878 0-5.447-4.431-9.879-9.878-9.879m0 21.057c-6.164 0-11.179-5.015-11.179-11.178 0-6.164 5.015-11.179 11.179-11.179 6.164 0 11.178 5.015 11.178 11.18 0 6.162-5.014 11.177-11.178 11.177"
          />
          <path
            d="M16.287 12.44h-.005a3.418 3.418 0 00-2.435 1.013 3.435 3.435 0 00.557 5.306 11.19 11.19 0 014.936-4.43 3.407 3.407 0 00-3.053-1.89zm-1.33 8.03l-.556-.241a4.74 4.74 0 01-2.859-4.336 4.707 4.707 0 011.384-3.357 4.715 4.715 0 013.354-1.396h.007c2.08 0 3.892 1.344 4.513 3.346l.18.579-.565.219a9.885 9.885 0 00-5.178 4.649l-.28.538zm13.704-6.14a11.179 11.179 0 014.935 4.428 3.438 3.438 0 001.563-2.879 3.45 3.45 0 00-3.446-3.44h-.006a3.407 3.407 0 00-3.046 1.89zm4.382 6.14l-.28-.537a9.884 9.884 0 00-5.178-4.649l-.563-.22.179-.578a4.713 4.713 0 014.504-3.346h.008a4.75 4.75 0 014.745 4.737 4.738 4.738 0 01-2.86 4.35l-.555.242z"
          />
        </g>
      </svg>
      <span
        class="tile__title"
      >
        Title
      </span>
    </div>
    <p
      class="tile__description"
    >
      Description
    </p>
  </a>
</div>
`;
